### DISCUSSION TITLE
Generate Random Strings Matching Predefined Pattern or Format for Enhanced Flexibility and Control

### DISCUSSION THREAD

**Discussion by wuletawwonte** [Dagic-zewdu/random-string#55] [Ideas]
Right now, it is not possible to get a random string based on a pattern or format provided. For example, provide "A0" as a pattern where A represents an uppercase letter and 0 represents a digit, and get something like "H7"

It would be nice if the package provides a method like `generateWithPattern(pattern)` that takes a pattern as an input and generates a random string based on that pattern. One possible approach could be adding pattern as an argument in the options object. What do you think about this feature?

**Comment by Dagic-zewdu** [comment]
What is your suggestion in implementation.

**Reply by wuletawwonte** [reply]
We may need to refactor it but we can do something like this
```ts

const characterSets: { [key: string]: string } = {
  a: 'abcdefghijklmnopqrstuvwxyz',
  A: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ',
  '0': '0123456789',
  '!': '~`!@#$%^&*()_+-={}[]:";\'<>?,./|\\',
};

const generateWithPattern = (pattern: string) => {
  let result = '';
  const patternLength = pattern.length;
  for (let i = 0; i < patternLength; i += 1) {
    const charSet = characterSets[pattern.charAt(i)];
    if(charSet) {
      const randomBytes = crypto.randomBytes(1);
      const randomIndex = Math.floor((randomBytes[0] / 256) * patternLength);
      result += charSet.charAt(randomIndex);
    } else {
      result += pattern.charAt(i);
    }
  }
  return result;
};

```
For a pattern like `Aa0!hN` this would generate a string similar to `Uw7*hN`. To integrate to the package we can get the charset from the `types.ts` file and even use the  `compute` method to avoid code duplication.
